src/jenkins_job_insight/storage.py

```
@@ -527,19 +531,28 @@ async def save_result(
         )
+        # Update the row (handles both fresh inserts and existing rows).
+        set_parts = [
+            "jenkins_url = COALESCE(NULLIF(?, ''), jenkins_url)",
+            "status = ?",
+            "result_json = ?",
+        ]
+        params: list = [jenkins_url, status, result_json]
+        if status == "completed":
+            set_parts.append("completed_at = COALESCE(completed_at, CURRENT_TIMESTAMP)")
+        params.append(job_id)
+        sql = f"UPDATE results SET {', '.join(set_parts)} WHERE job_id = ?"
```

Ruff (0.15.7)

[error] Possible SQL injection vector through string-based query construction

(S608)